src/set_vault_jwt_expiration/set_vault_jwt_expiration/set.py

```diff
@@ -0,0 +1,64 @@
+import logging
+
+import datetime
+import jwt
+
+import set_vault_jwt_expiration.context
+
+from lib.jwt import is_jwt
+from lib.util import (
+  url_host_port
+)
+from lib.vault import Vault
+
+
+def set_jwt_expiration(
+  artsy_env,
+  artsy_project,
+  vault_host,
+  vault_port,
+  kvv2_mount_point,
+):
+  ''' set expires_at custom header for all JWTs '''
+  vault_path = 'kubernetes/apps/' + f'{artsy_project}/'
+  vault_client = Vault(
+    url_host_port(vault_host, vault_port),
+    auth_method='iam',
+    kvv2_mount_point=kvv2_mount_point,
+    path=vault_path,
+  )
+  # get list of jwt vars
+  jwt_vars = vault_client.list(match_function=is_jwt, match_type='value')
+  logging.info(f'List of JWTs: {jwt_vars}')
+  for var in jwt_vars:
+    logging.info(
+      f"Checking JWT {var}..."
+    )
+    jwt_string = vault_client.get(var)
+    set_exp_meta(vault_client, var, jwt_string)
+
+def set_exp_meta(vault_client, var, jwt_str):
+  ''' extract expiration date from jwt payload and set it in vault custom metadata '''
+  exp_date_filler = 'nil'
+  exp_date_meta_key = 'expires_at'
+  payload = jwt.decode(
+    jwt_str,
+    options={"verify_signature": False, "verify_exp": False}
+  )
+  if 'exp' in payload and payload['exp'] is not None:
+    exp_payload = payload['exp']
+    logging.info(f"JWT has exp payload: {exp_payload}")
+    exp_date = datetime.datetime.utcfromtimestamp(exp_payload)
+    logging.info(f"Exp payload to date: {exp_date}")
+    exp_date_str = exp_date.isoformat("T") + "Z"
+  else:
+    logging.info(
+      f"JWT does not have exp payload or exp payload is not valid"
+    )
+    exp_date_str = exp_date_filler
```

Comment:
Your call, but wonder if even a empty string would be better, if supported? But string of `nil` feels bound for some confusion.

Reply:
i am fine with either. problem is, if blank string, i am not sure what value hvac returns to the script (null string? None? ...)

Reply:
@cavvia - wonder what you think.

Reply:
As a client I would expect either an empty string or a `None` type object back from a Vault python client if a null value is set for a key... either of those can easily be tested with a python conditional (empty strings return false). 

It seems we are using KV v1 to read the secrets in Fortress [here](https://github.com/artsy/fortress/blob/main/src/fortress/lib/vault.py#L50). That seems to use JSON serialization in the hvac client [here](https://github.com/hvac/hvac/blob/main/hvac/api/secrets_engines/kv_v1.py#L26). JSON does support `null` values and python will convert those to None objects, e.g.:

```
>>> import json
>>> json.loads('{"my_key": null}')
{'my_key': None}
```

So the question of whether Vault will return None or "" is based on whether vault is using JSON `null` values under the hood. If so this will work end to end with `None` values.

Reply:
tested setting to empty string and `None`. hvac throws:
```
hvac.exceptions.InvalidRequest: 1 error occurred:
        * custom_metadata validation failed: length of value for key "expires_at" is 0 but must be 0 < len(value) <= 512
```

also tested setting to `False`, vault cli shows `0`, and hvac reads it as `0` string which is truthy.
```
artsy-2:~ jxu$ vault kv get -mount kvv2 kubernetes/apps/hokusai-sandbox/TOKEN
================= Secret Path =================
kvv2/data/kubernetes/apps/hokusai-sandbox/TOKEN

======= Metadata =======
Key                Value
---                -----
created_time       2024-12-16T16:25:59.457131248Z
custom_metadata    map[expires_at:0]
deletion_time      n/a
destroyed          false
version            2

-> for var in jwt_vars:
(Pdb) custom_meta
{'expires_at': '0'}
(Pdb) val = custom_meta['expires_at']
(Pdb) val
'0'
(Pdb) type(val)
<class 'str'>
```

Reply:
so let's stick with 'nil' string.